processor.go

```diff
@@ -442,11 +443,133 @@ func callInterfaceFunc(handler, input interface{}, streaming bool) (intf interfa
 	return
 }
 
+// ProcessClientStreamRequest allows clients to push multiple messages
+// asynchronously to the same service with the same connection. Unlike in
+// ProcessClientRequest, we take a channel of inputs that can be filled and
+// will subsequently call the service with any new messages received in the
+// channel. The caller is responsible for closing the client input channel when
+// it is done.
+func (p *ServiceProcessor) ProcessClientStreamRequest(req *http.Request, path string,
+	clientInputs chan []byte) (chan []byte, error) {
+
+	outChan := make(chan []byte, 100)
+	var closeOutOnce sync.Once
+	mh, ok := p.handlers[path]
+
+	if !ok {
+		err := xerrors.New("the requested message hasn't been " +
+			"registered: " + path)
+		log.Error(err)
+		return nil, err
+	}
+
+	var stopServiceChan chan bool
+	var reply interface{}
+
+	// This goroutine listens on any new messages from the client and executes
+	// the request. Executing the request should fill the service's channel, as
+	// the service will use the same chanel for further requests.
+	go func() {
+		for {
+			select {
+			case buf, ok := <-clientInputs:
+				if !ok {
+					if stopServiceChan != nil {
+						close(stopServiceChan)
+					}
+					return
+				}
+
+				msg := reflect.New(mh.msgType).Interface()
+
+				err := protobuf.DecodeWithConstructors(buf, msg,
+					network.DefaultConstructors(p.Context.server.Suite()))
+				if err != nil {
+					log.Error(xerrors.Errorf("decoding: %v", err))
+					return
+				}
+
+				reply, stopServiceChan, err = callInterfaceFunc(mh.handler, msg, mh.streaming)
+				if err != nil {
+					log.Error(err)
+					if stopServiceChan != nil {
+						close(stopServiceChan)
+					}
+				}
+
+				// This goroutine is responsible for listening on the service channel,
+				// decoding the messages and then forwarding them to the streaming
+				// tunnel, which should then forward the message to the client. A new
+				// routine is created each time the client makes a request.
+				go func() {
+					inChan := reflect.ValueOf(reply)
+					cases := []reflect.SelectCase{
+						reflect.SelectCase{Dir: reflect.SelectRecv, Chan: inChan},
+					}
+
+					// Since this goroutine is created each time the client sends a
+					// request, we then must ensure the outChan is closed only once.
+					defer func() {
+						closeOutOnce.Do(func() {
+							close(outChan)
+						})
+					}()
+
+					for {
+						chosen, v, ok := reflect.Select(cases)
+						if !ok {
+							log.Lvlf4("publisher is closed for %s, closing "+
+								"outgoing channel", path)
+							return
+						}
+						if chosen == 0 {
+							// Send information down to the client.
+							buf, err = protobuf.Encode(v.Interface())
+							if err != nil {
+								log.Error(err)
+								return
+							}
+							outChan <- buf
```

Comment:
I think you could end up in a situation where two go routines are running, one closes the out channel and the second tries to write into it which will end up in a panic.